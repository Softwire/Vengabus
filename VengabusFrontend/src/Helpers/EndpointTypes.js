var EndpointTypes = Object.freeze({
    QUEUE: "queue",
    TOPIC: "topic",
    SUBSCRIPTION: "subscription",
    MESSAGE: "message",
    DEADLETTER: "deadletters"
});


var typeToTitle = (type) => {
    switch (type) {
        case EndpointTypes.MESSAGE:
            return "Messages";
        case EndpointTypes.SUBSCRIPTION:
            return "Subscriptions";
        case EndpointTypes.QUEUE:
            return "Queues";
        case EndpointTypes.TOPIC:
            return "Topics";
        case EndpointTypes.DEADLETTER:
            return "Deadletters";
        default:
            throw new Error('Invalid endpoint type.');
    }
}
export { EndpointTypes, typeToTitle };