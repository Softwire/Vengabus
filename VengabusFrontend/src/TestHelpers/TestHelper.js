import { Button } from "react-bootstrap";

class TestHelper {
    /*
     * It returns a promise that waits until the state of the components is updated and then continues.
     * 
     * Expected usage is: 
     *     return testHelper.afterReactHasUpdated().then(() => {
     *         expect('actualValue').toEqual('expectedValue');
     *     });
     * 
     * Notes on multi-layered async tests (e.g. wait for this, and then wait for that, and then expect a result) will be forthcoming.
     */
    afterReactHasUpdated(noArgsGiven) {
        if (noArgsGiven) {
            throw new Error('afterReactHasUpdated() does not expect to take any arguments. See expected usage for further details.');
        }

        return new Promise(function (resolve) {
            setTimeout(resolve, 0);
        });
    }
    clickElementWithId(wrapper, id) {
        wrapper.find(id).at(0).simulate("click");
    }
}

export const testHelper = new TestHelper();
