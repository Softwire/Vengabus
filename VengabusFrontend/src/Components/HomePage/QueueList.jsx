import React, { Component } from 'react';
import { DataTable } from '../DataTable';
import { formatTimeStamp } from '../../Helpers/FormattingHelpers';
import { palerBlue } from '../../colourScheme';

export class QueueList extends Component {

    render() {
        let queueArray = undefined;
        if (this.props.queueData) {
            queueArray = [...this.props.queueData];

            for (let i = 0; i < queueArray.length; i++) {
                //needs to be cloned
                queueArray[i] = { ...queueArray[i] };
                const currentMessageArray = queueArray[i];
                if (currentMessageArray.mostRecentDeadLetter) {
                    currentMessageArray.mostRecentDeadLetter = formatTimeStamp(currentMessageArray.mostRecentDeadLetter);
                }
            }
        }

        const colProps = [
            {
                dataField: 'name',
                width: 25,
            },
            {
                dataField: 'activeMessageCount',
                width: 25,
                align: 'right'
            },
            {
                dataField: 'deadletterMessageCount',
                width: 25,
                align: 'right'
            },
            {
                dataField: 'mostRecentDeadLetter',
                width: 25,
            }
        ];

        const selectRow = {
            bgColor: palerBlue,
            selected: this.props.currentlySelectedName
        };

        return (
            <DataTable
                name='QueueList'
                colProps={colProps}
                dataToDisplay={queueArray}
                defaultHover
                onRowClick={this.props.clickFunction}
                selectRow={selectRow}
                uniqueKeyColumn='name'
                searchable
            />
        );
    }
}
