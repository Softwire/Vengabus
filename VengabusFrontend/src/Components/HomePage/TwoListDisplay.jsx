import React, { Component } from 'react';
import { serviceBusConnection } from '../../AzureWrappers/ServiceBusConnection';
import { QueueList } from './QueueList';
import { TopicList } from './TopicList';
import { MessageList } from './MessageList';
import { css } from 'react-emotion';
import { Breadcrumb, Button } from 'react-bootstrap';
import { SubscriptionList } from './SubscriptionList';
import { EndpointTypes, typeToTitle } from '../../Helpers/EndpointTypes';
import { sharedSizesAndDimensions } from '../../Helpers/SharedSizesAndDimensions';
import { cancellablePromiseCollection } from '../../Helpers/CancellablePromiseCollection';

const messageCount = 500;

export class TwoListDisplay extends Component {
    constructor(props) {
        super(props);
        this.displayHistory = [];
        this.breadCrumbHistory = [{ name: "Home", type: undefined }];
        this.messageButtonDisabled = false;
        this.promiseCollection = new cancellablePromiseCollection();
        this.state = {
            queueData: undefined,
            topicData: undefined,
            subscriptionData: undefined,
            messageData: undefined,
            rightTableType: EndpointTypes.TOPIC
        };
    }

    componentDidMount() {
        serviceBusConnection.registerForUpdatesPrompts(this.resetInitialStateData);

    }
    componentWillUnmount() {
        serviceBusConnection.deregisterForUpdatesPrompts(this.resetInitialStateData);
        this.promiseCollection.cancelAllPromises();

    }

    handleQueueRowClick = (e, row, rowIndex) => {
        this.breadCrumbHistory = [{ name: "Home", type: undefined }, { name: row.name, type: EndpointTypes.QUEUE }];
        this.messageButtonDisabled = true;
        this.setState({
            messageData: undefined,
            rightTableType: EndpointTypes.MESSAGE
        }, this.updateEndpointMessageData);
    }

    handleTopicRowClick = (e, row, rowIndex) => {
        this.breadCrumbHistory = [{ name: "Home", type: undefined }, { name: row.name, type: EndpointTypes.TOPIC }];
        this.setState({
            subscriptionData: undefined,
            rightTableType: EndpointTypes.SUBSCRIPTION
        }, this.updateTopicSubscriptionData);
    }

    handleSubscriptionRowClick = (e, row, rowIndex) => {
        this.breadCrumbHistory[2] = { name: row.name, type: EndpointTypes.SUBSCRIPTION };
        this.messageButtonDisabled = true;
        this.setState({
            messageData: undefined,
            rightTableType: EndpointTypes.MESSAGE
        }, () => this.updateEndpointMessageData(false));
    }

    handleMessageToggle = (isNewTypeDeadLetter) => {
        this.messageButtonDisabled = true;
        this.setState({
            messageData: undefined,
            rightTableType: isNewTypeDeadLetter ? EndpointTypes.DEADLETTER : EndpointTypes.MESSAGE
        }, () => this.updateEndpointMessageData(isNewTypeDeadLetter));
    }

    updateAllQueueData = () => {
        const serviceBusService = serviceBusConnection.getServiceBusService();
        this.promiseCollection.cancelAllPromises(EndpointTypes.QUEUE);
        const fetchedQueueData = this.promiseCollection.addNewPromise(serviceBusService.listQueues(), EndpointTypes.QUEUE);
        fetchedQueueData.then((result) => {
            this.setState({
                queueData: result.map((item) => ({ ...item, mostRecentDeadLetter: null }))
            });
            let currentQueueData = result;
            for (let i = 0; i < result.length; i++) {
                const getMostRecentDLPromise = this.promiseCollection.addNewPromise(serviceBusService.getMostRecentDeadletter(result[i].name), result[i].name);
                getMostRecentDLPromise.then((timeResult) => {
                    const mostRecentDeadLetterTimestamp = timeResult;
                    currentQueueData[i].mostRecentDeadLetter = mostRecentDeadLetterTimestamp;
                    this.setState({
                        queueData: currentQueueData
                    });
                }).catch((e) => { });
            }
        }).catch((e) => { });
    }

    updateAllTopicData = () => {
        const serviceBusService = serviceBusConnection.getServiceBusService();
        this.promiseCollection.cancelAllPromises(EndpointTypes.TOPIC);
        const fetchedTopicData = this.promiseCollection.addNewPromise(serviceBusService.listTopics(), EndpointTypes.TOPIC);
        fetchedTopicData.then(result => {
            this.setState({
                topicData: result
            });
        }).catch(e => { if (!e.isCanceled) { console.log(e); } });
    }

    updateTopicSubscriptionData = () => {
        const topicName = this.breadCrumbHistory[this.breadCrumbHistory.length - 1].name;
        const serviceBusService = serviceBusConnection.getServiceBusService();
        this.promiseCollection.cancelAllPromises(EndpointTypes.SUBSCRIPTION);
        const fetchedSubscriptionData = this.promiseCollection.addNewPromise(serviceBusService.listSubscriptions(topicName), EndpointTypes.SUBSCRIPTION);
        fetchedSubscriptionData.then(result => {
            this.setState({
                subscriptionData: result
            });
        }).catch(e => { if (!e.isCanceled) { console.log(e); } });
    }


    updateEndpointMessageData = (isMessageDeadletters) => {
        const serviceBusService = serviceBusConnection.getServiceBusService();
        let fetchedMessageData;
        const leftTableType = this.breadCrumbHistory[this.breadCrumbHistory.length - 1].type;
        if (leftTableType === EndpointTypes.QUEUE || typeof leftTableType === 'undefined') {
            const queueName = this.breadCrumbHistory[this.breadCrumbHistory.length - 1].name;
            if (isMessageDeadletters) {
                fetchedMessageData = serviceBusService.listQueueDeadLetterMessages(queueName, messageCount);
            } else {
                fetchedMessageData = serviceBusService.listQueueMessages(queueName, messageCount);
            }
        } else {
            const topicName = this.breadCrumbHistory[this.breadCrumbHistory.length - 2].name;
            const subscriptionName = this.breadCrumbHistory[this.breadCrumbHistory.length - 1].name;
            if (isMessageDeadletters) {
                fetchedMessageData = serviceBusService.listSubscriptionDeadLetterMessages(topicName, subscriptionName, messageCount);
            } else {
                fetchedMessageData = serviceBusService.listSubscriptionMessages(topicName, subscriptionName, messageCount);
            }
        }
        this.promiseCollection.cancelAllPromises(EndpointTypes.MESSAGE);
        const wrappedFetchedMessageData = this.promiseCollection.addNewPromise(fetchedMessageData, EndpointTypes.MESSAGE);
        wrappedFetchedMessageData.then((result) => {
            this.messageButtonDisabled = false;
            this.setState({
                messageData: result
            });
        }).catch(e => { if (!e.isCanceled) { console.log(e); } });
    }

    resetInitialStateData = () => {
        this.updateAllTopicData();
        this.updateAllQueueData();
        this.breadCrumbHistory = [{ name: "Home", type: undefined }];
        this.setState({
            rightTableType: EndpointTypes.TOPIC
        });
    };

    getDeadLetterToggleButtonText = (isDeadLetterMessage) => {
        return isDeadLetterMessage ? "Deadletters" : "Live Messages";
    }

    getList = (isForRightHandList) => {
        let typeOfData;
        const currentLeftTable = this.breadCrumbHistory[this.breadCrumbHistory.length - 1];
        let currentSelection;
        if (isForRightHandList) {
            typeOfData = this.state.rightTableType;
        } else {
            //if there is no history the currentLeftTable will be undefined
            //if its undefined its undefined then its the original state and therefore should be queue
            typeOfData = currentLeftTable.type || EndpointTypes.QUEUE;
            currentSelection = currentLeftTable.name;
        }
        const displayStyle = css`
                    display: inline-block;
                    margin:10px;
                `;

        const deadLetterToggleButtonStyle = css`
                margin-right: 0px;
                float: right;
                margin: 9px;
                width: 120px ;
        `;
        const minHeightOfHeader = {
            "minHeight": "92px",
            "height": "92px"
        };

        switch (typeOfData) {
            case EndpointTypes.QUEUE:
                return (
                    <React.Fragment>
                        <div>
                            <h2 id='title'>{typeToTitle(EndpointTypes.QUEUE)}</h2>
                        </div>
                        <QueueList
                            queueData={this.state.queueData}
                            clickFunction={this.handleQueueRowClick}
                            currentlySelectedName={currentSelection}
                            id='QueueTable'
                            headerStyle={minHeightOfHeader}
                        />
                    </React.Fragment>
                );
            case EndpointTypes.TOPIC:
                return (
                    <React.Fragment>
                        <div>
                            <h2 id='title'>{typeToTitle(EndpointTypes.TOPIC)}</h2>
                        </div>
                        <TopicList
                            topicData={this.state.topicData}
                            clickFunction={this.handleTopicRowClick}
                            currentlySelectedName={currentSelection}
                            id='TopicTable'
                            headerStyle={minHeightOfHeader}
                        />
                    </React.Fragment>
                );

            case EndpointTypes.SUBSCRIPTION:
                return (
                    <React.Fragment>
                        <div>
                            <h2 id='title' >{typeToTitle(EndpointTypes.SUBSCRIPTION)}</h2>
                        </div>
                        <SubscriptionList
                            subscriptionData={this.state.subscriptionData}
                            clickFunction={this.handleSubscriptionRowClick}
                            currentlySelectedName={currentSelection}
                            id='SubscriptionTable'
                            headerStyle={minHeightOfHeader}
                        />
                    </React.Fragment>
                );
            case EndpointTypes.MESSAGE:
            case EndpointTypes.DEADLETTER:
                const isDeadLetterMessage = typeOfData === EndpointTypes.DEADLETTER;
                const lastBreadCrumb = this.breadCrumbHistory[this.breadCrumbHistory.length - 1];
                const penultimateBreadCrumb = this.breadCrumbHistory[this.breadCrumbHistory.length - 2];
                return (
                    <React.Fragment>
                        <div >
                            <h2 className={displayStyle} id='title'>{typeToTitle(typeOfData)}</h2>
                            <Button className={deadLetterToggleButtonStyle} onClick={() => this.handleMessageToggle(!isDeadLetterMessage)} disabled={this.messageButtonDisabled} >
                                {this.getDeadLetterToggleButtonText(!isDeadLetterMessage)}
                            </Button>
                        </div>
                        <MessageList
                            id='MessageTable'
                            messageData={this.state.messageData}
                            messageType={typeOfData}
                            endpointType={lastBreadCrumb.type}
                            endpointName={lastBreadCrumb.name}
                            endpointParent={penultimateBreadCrumb.name}
                            headerStyle={minHeightOfHeader}
                            refreshMessageTableHandler={() => {
                                this.setState({
                                    messageData: undefined
                                }, () => this.updateEndpointMessageData(isDeadLetterMessage));

                            }}
                        />
                    </React.Fragment>

                );
            default:
                throw new Error('Invalid endpoint type.');
        }
    }

    HandleBreadCrumbClick = (type, newLength) => {
        switch (type) {
            case (EndpointTypes.QUEUE):
                this.breadCrumbHistory = this.breadCrumbHistory.slice(0, newLength + 1);
                this.updateAllQueueData();
                this.updateEndpointMessageData();
                this.setState({
                    rightTableType: EndpointTypes.MESSAGE
                });
                break;
            case (EndpointTypes.TOPIC):
                this.breadCrumbHistory = this.breadCrumbHistory.slice(0, newLength + 1);
                this.updateAllTopicData();
                this.updateTopicSubscriptionData();
                this.setState({
                    rightTableType: EndpointTypes.SUBSCRIPTION
                });
                break;
            case (EndpointTypes.SUBSCRIPTION):
                this.breadCrumbHistory = this.breadCrumbHistory.slice(0, newLength + 1);
                this.updateTopicSubscriptionData();
                this.updateEndpointMessageData();
                this.setState({
                    rightTableType: EndpointTypes.MESSAGE
                });
                break;
            default:
                this.breadCrumbHistory = this.breadCrumbHistory.slice(0, newLength + 1);
                this.resetInitialStateData();
                break;
        }

    }


    getBreadcrumbElement = () => {
        const breadCrumbStyle = css`
        &.breadcrumb{
          float: left;
            height:${sharedSizesAndDimensions.BREADCRUMBS_HEIGHT}px;
            margin-bottom: 1px;
        }
        `;

        const breadcrumbItems = this.breadCrumbHistory.map((breadCrumb, i) => {
            return (<Breadcrumb.Item id={this.breadCrumbHistory[i].name} onClick={() => this.HandleBreadCrumbClick(breadCrumb.type, i)} active={(i === this.breadCrumbHistory.length - 1)} key={i}>
                {this.breadCrumbHistory[i].name}
            </Breadcrumb.Item>);
        });
        const areOnHomePage = (this.breadCrumbHistory.length === 1);

        if (!areOnHomePage) {
            breadcrumbItems.push(
                (<Breadcrumb.Item key={breadcrumbItems.length} active>
                    {typeToTitle(this.state.rightTableType)}
                </Breadcrumb.Item>)
            );
        }

        return (
            <Breadcrumb className={breadCrumbStyle}>
                {breadcrumbItems}
            </Breadcrumb>
        );
    }

    render() {
        const tableMarginPixels = 15;

        const displayStyle = css`
            width: calc(50% - ${tableMarginPixels * 2}px);
            margin-left:${tableMarginPixels}px;
            margin-right:${tableMarginPixels}px;
            display: inline-block; /*to allow tables to be displayed side by side*/
        `;

        const outerDivDisplay = css`
            display: block;
            `;

        const breadCrumbDisplay = css`
            display: block;
            height:${sharedSizesAndDimensions.BREADCRUMBS_HEIGHT}px;
            margin:2px;
            `;

        const minWidth = css`
            min-width:1000px;
        `;

        const areOnHomePage = (this.breadCrumbHistory.length === 1);
        const totalDiff = sharedSizesAndDimensions.DEFAULT_HEADER_HEIGHT + sharedSizesAndDimensions.BREADCRUMBS_HEIGHT + sharedSizesAndDimensions.BOTTOM_GUTTERING;

        const line = css`
            border-left: 1px solid black;
            display : ${(areOnHomePage) ? "inline-block" : "none"};
            height : calc(100% - ${totalDiff}px); 
            position: fixed;
            z-index: -1;
        `;

        const leftBox = this.getList();
        const rightBox = this.getList(true);



        return (
            <div className={minWidth} >
                <div className={breadCrumbDisplay} >
                    {this.getBreadcrumbElement()}
                </div>
                <div className={outerDivDisplay}>
                    <div className={displayStyle} id="left">
                        {leftBox}
                    </div>
                    <div className={line} />
                    <div className={displayStyle} id="right">
                        {rightBox}
                    </div>
                </div>
            </div>
        );
    }
}

