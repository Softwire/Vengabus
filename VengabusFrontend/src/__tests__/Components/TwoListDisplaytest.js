import { mount } from 'enzyme';
import renderer from 'react-test-renderer';
import React from 'react';
import { TwoListDisplay } from "../../Components/HomePage/TwoListDisplay";
import { testHelper } from '../../TestHelpers/TestHelper';
import toJson from 'enzyme-to-json';
import { EndpointTypes } from '../../Helpers/EndpointTypes';

//event object to hand to click functions
const eventObj = {};

jest.mock('../../AzureWrappers/VengaServiceBusService', () => ({
    VengaServiceBusService: class {
        constructor() {
        }

        listTopics = () => {
            return Promise.resolve(
                [
                    {
                        name: "testTopic1",
                        subscriptionCount: 12,
                        topicStatus: "Active"
                    },
                    {
                        name: "testTopic2",
                        subscriptionCount: 13,
                        topicStatus: "Active"
                    },
                    {
                        name: "testTopic3",
                        subscriptionCount: 13,
                        topicStatus: "Active"
                    }
                ]
            );
        }

        listQueues = () => {
            return Promise.resolve([
                {
                    name: "testQueue1",
                    activeMessageCount: 12,
                    deadletterMessageCount: 17
                },
                {
                    name: "testQueue2",
                    activeMessageCount: 13,
                    deadletterMessageCount: 15
                },
                {
                    name: "testQueue3",
                    activeMessageCount: 13,
                    deadletterMessageCount: 18
                }
            ]
            );
        }

        listQueueMessages = (queueName) => {
            return (queueName === "testQueue1") ? Promise.resolve(
                [
                    {
                        predefinedProperties: { messageId: "test1" },
                        uniqueId: "59298c2b-d58f-4ad0-bde9-f8a9d00a3070",
                        messageBody: "apple"

                    },
                    {
                        predefinedProperties: { messageId: "test2" },
                        uniqueId: "c9f547bf-72e1-439e-bd1f-0b590422a6f8",
                        messageBody: "banana"
                    },
                    {
                        predefinedProperties: { messageId: "test3" },
                        uniqueId: "5034e2f8-9bf0-436a-b8f4-914b43594ee1",
                        messageBody: "carrot"
                    }
                ]
            ) : Promise.reject();
        }

        listSubscriptions = (topicName) => {
            return (topicName === "testTopic1") ? Promise.resolve([
                {
                    name: "testSubscriptions1",
                    activeMessageCount: 12,
                    deadletterMessageCount: 17
                },
                {
                    name: "testSubscriptions2",
                    activeMessageCount: 13,
                    deadletterMessageCount: 15
                },
                {
                    name: "testSubscriptions3",
                    activeMessageCount: 13,
                    deadletterMessageCount: 18
                }
            ]
            ) : Promise.reject();

        }

        listSubscriptionMessages = () => {
            return Promise.resolve(
                [
                    {
                        predefinedProperties: { messageId: "test1" },
                        uniqueId: "57d7d4dd-291a-453c-a0e4-efbb664607c0",
                        messageBody: "apple"

                    },
                    {
                        predefinedProperties: { messageId: "test2" },
                        uniqueId: "873c61c3-797a-4158-941b-da8eb7410e70",
                        messageBody: "banana"
                    },
                    {
                        predefinedProperties: { messageId: "test3" },
                        uniqueId: "8ab56d4c-0204-4aa2-a888-2cc305cd0275",
                        messageBody: "carrot"
                    }
                ]);
        }
    }
}));

it('renders correctly before data is added', () => {
    let messagePropertyInput = renderer.create(
        <TwoListDisplay />);
    expect(messagePropertyInput.toJSON()).toMatchSnapshot();
});

it('renders correctly after data is added', () => {
    let wrapper = mount(<TwoListDisplay />);
    return resetToDefaultState(wrapper).then(
        () => { expect(toJson(wrapper)).toMatchSnapshot(); }
    );
});

it('renders queues and topic titles', () => {
    let wrapper = mount(<TwoListDisplay />);

    const leftTitle = wrapper.find('#left').find('#title').text();
    const rightTitle = wrapper.find('#right').find('#title').text();
    expect(leftTitle).toBe("Queues");
    expect(rightTitle).toBe("Topics");
});

it('queues and topics populate', () => {
    let wrapper = mount(<TwoListDisplay />);

    return resetToDefaultState(wrapper).then(
        () => { return expectDefaultState(wrapper); }
    );
});

it('clicking Queues retrieves messages', () => {
    let wrapper = mount(<TwoListDisplay />);

    return resetToDefaultState(wrapper).then(
        () => { return clickAndResolveQueueRowClick(wrapper, "testQueue1"); }).then(
            () => { return expectsQueueAndMessage(wrapper); }
        );
});

it('clicking Queues to retrieved message data then returning to home screen', () => {
    let wrapper = mount(<TwoListDisplay />);

    return resetToDefaultState(wrapper).then(
        () => { return clickAndResolveQueueRowClick(wrapper, "testQueue1"); }).then(
            () => { return resetToDefaultState(wrapper); }).then(
                () => { return expectDefaultState(wrapper); }
            );
});

it('clicking Topics retrieves subscriptions', () => {
    let wrapper = mount(<TwoListDisplay />);
    return resetToDefaultState(wrapper).then(
        () => { return clickAndResolveTopicRowClick(wrapper, "testTopic1"); }).then(
            () => { return expectSubscriptionsAndTopics(wrapper); }
        );
});

it('clicking topics then Subscriptions retrieves messages', () => {
    let wrapper = mount(<TwoListDisplay />);

    return resetToDefaultState(wrapper).then(
        () => { return clickAndResolveTopicRowClick(wrapper, "testTopic1"); }).then(
            () => { return clickAndResolveSubscriptionRowClick(wrapper, "testSubscriptions1"); }).then(
                () => { return expectSubscriptionAndMessage(wrapper); }
            );
});

it('Home breadCrumb resets state', () => {
    let wrapper = mount(<TwoListDisplay />);
    return afterBreadCrumbClick(undefined, wrapper);

});

it('Can go back to Topic from Subscription using BreadCrumbs', () => {
    let wrapper = mount(<TwoListDisplay />);
    return afterBreadCrumbClick(EndpointTypes.TOPIC, wrapper);
});


//Helper functions used for repeating tests
const resetToDefaultState = (wrapper) => {
    wrapper.update();
    wrapper.instance().resetInitialStateData();
    return testHelper.afterReactHasUpdated();
};

const clickAndResolveQueueRowClick = (wrapper, queueName) => {
    wrapper.update();
    const queueList = wrapper.find('#QueueTable');
    queueList.props().clickFunction(eventObj, { name: queueName });

    return testHelper.afterReactHasUpdated();
};

const clickAndResolveTopicRowClick = (wrapper, topicName) => {
    wrapper.update();
    const topic = wrapper.find('#TopicTable');
    topic.props().clickFunction(eventObj, { name: topicName });
    return testHelper.afterReactHasUpdated();
};

const clickAndResolveSubscriptionRowClick = (wrapper, subscriptionName) => {
    wrapper.update();
    const subscriptionTable = wrapper.find('#SubscriptionTable');
    subscriptionTable.props().clickFunction(eventObj, { name: subscriptionName });

    return testHelper.afterReactHasUpdated();
};


const afterBreadCrumbClick = (breadcrumb, wrapper) => {
    switch (breadcrumb) {
        case (EndpointTypes.TOPIC):
            return clickAndResolveTopicRowClick(wrapper, "testTopic1").then(
                () => { return clickAndResolveSubscriptionRowClick(wrapper, "testSubscriptions1"); }).then(
                    () => {
                        wrapper.update();
                        const topicBreadCrumb = wrapper.find('#testTopic1').hostNodes();
                        topicBreadCrumb.simulate("click");
                        return testHelper.afterReactHasUpdated();
                    }).then(
                        () => {
                            return expectSubscriptionsAndTopics(wrapper);
                        });
        default:
            return clickAndResolveQueueRowClick(wrapper, "testQueue1").then(
                () => {
                    wrapper.update();
                    const homeBreadCrumb = wrapper.find('#Home').hostNodes();
                    homeBreadCrumb.simulate("click");
                    return testHelper.afterReactHasUpdated();
                }).then(
                    () => {
                        return expectDefaultState(wrapper);
                    });
    }
};

const expectDefaultState = (wrapper) => {
    wrapper.update();

    const queueTable = wrapper.find('#QueueTable').find("#Data");
    const topicTable = wrapper.find('#TopicTable').find("#Data");

    const leftTitle = wrapper.find('#left').find('#title').text();
    const rightTitle = wrapper.find('#right').find('#title').text();

    expect(leftTitle).toBe("Queues");
    expect(rightTitle).toBe("Topics");

    expect(queueTable).toExistOnPage();
    expect(topicTable).toExistOnPage();
};

const expectSubscriptionsAndTopics = (wrapper) => {
    wrapper.update();

    const topicTable = wrapper.find('#TopicTable').find("#Data")
    const subscriptionTable = wrapper.find('#SubscriptionTable').find("#Data");

    const leftTitle = wrapper.find('#left').find('#title').text();
    const rightTitle = wrapper.find('#right').find('#title').text();

    expect(leftTitle).toBe("Topics");
    expect(rightTitle).toBe("Subscriptions");

    expect(topicTable).toExistOnPage();
    expect(subscriptionTable).toExistOnPage();
};

const expectsQueueAndMessage = (wrapper) => {
    wrapper.update();

    const queueTable = wrapper.find('#QueueTable').find("#Data");
    const messageTable = wrapper.find('#MessageTable').find("#Data");

    const leftTitle = wrapper.find('#left').find('#title').text();
    const rightTitle = wrapper.find('#right').find('#title').text();

    expect(leftTitle).toBe("Queues");
    expect(rightTitle).toBe("Messages");

    expect(queueTable).toExistOnPage();
    expect(messageTable).toExistOnPage();
};

const expectSubscriptionAndMessage = (wrapper) => {
    wrapper.update();

    const subscriptionTable = wrapper.find('#SubscriptionTable').find("#Data");
    const messageTable = wrapper.find('#MessageTable').find("#Data");

    const leftTitle = wrapper.find('#left').find('#title').text();
    const rightTitle = wrapper.find('#right').find('#title').text();

    expect(leftTitle).toBe("Subscriptions");
    expect(rightTitle).toBe("Messages");

    expect(subscriptionTable).toExistOnPage();
    expect(messageTable).toExistOnPage();
};

