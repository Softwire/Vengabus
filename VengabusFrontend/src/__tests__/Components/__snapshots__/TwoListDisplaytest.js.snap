// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly after data is added 1`] = `
.emotion-4 {
  width: calc(50% - 30px);
  margin-left: 15px;
  margin-right: 15px;
  display: inline-block;
}

.emotion-8 {
  display: block;
}

.emotion-2 {
  display: block;
  height: 35px;
  margin: 2px;
}

.emotion-9 {
  min-width: 1000px;
}

.emotion-5 {
  border-left: 1px solid black;
  display: inline-block;
  height: calc(100% - 100px);
  position: fixed;
  z-index: -1;
}

.emotion-0.breadcrumb {
  float: left;
  height: 35px;
  margin-bottom: 1px;
}

.emotion-3 {
  text-align: center;
  float: left;
  width: 100%;
}

<TwoListDisplay>
  <div
    className="emotion-9"
  >
    <div
      className="emotion-2"
    >
      <Breadcrumb
        bsClass="breadcrumb"
        className="emotion-0"
      >
        <ol
          aria-label="breadcrumbs"
          className="emotion-0 breadcrumb"
          role="navigation"
        >
          <BreadcrumbItem
            active={true}
            id="Home"
            key="0"
            onClick={[Function]}
          >
            <li
              className="active"
            >
              <span
                id="Home"
                onClick={[Function]}
              >
                Home
              </span>
            </li>
          </BreadcrumbItem>
        </ol>
      </Breadcrumb>
    </div>
    <div
      className="emotion-8"
    >
      <div
        className="emotion-4"
        id="left"
      >
        <div>
          <h2
            id="title"
          >
            Queues
          </h2>
        </div>
        <QueueList
          clickFunction={[Function]}
          currentlySelectedName="Home"
          id="QueueTable"
        >
          <DataTable
            colProps={
              Array [
                Object {
                  "dataField": "name",
                  "headerClasses": "css-1lvmfbt",
                  "width": 25,
                },
                Object {
                  "align": "right",
                  "dataField": "activeMessageCount",
                  "headerClasses": "css-1lvmfbt",
                  "width": 25,
                },
                Object {
                  "align": "right",
                  "dataField": "deadletterMessageCount",
                  "headerClasses": "css-1lvmfbt",
                  "width": 25,
                },
                Object {
                  "dataField": "mostRecentDeadLetter",
                  "headerClasses": "css-1lvmfbt",
                  "width": 25,
                },
              ]
            }
            defaultHover={true}
            name="QueueList"
            onRowClick={[Function]}
            selectRow={
              Object {
                "bgColor": "#99ccff",
                "selected": "Home",
              }
            }
            uniqueKeyColumn="name"
          >
            <p
              className="emotion-3"
            >
              No data has been retrieved yet.
            </p>
          </DataTable>
        </QueueList>
      </div>
      <div
        className="emotion-5"
      />
      <div
        className="emotion-4"
        id="right"
      >
        <div>
          <h2
            id="title"
          >
            Topics
          </h2>
        </div>
        <TopicList
          clickFunction={[Function]}
          id="TopicTable"
        >
          <DataTable
            colProps={
              Array [
                Object {
                  "dataField": "name",
                  "headerClasses": "css-1lvmfbt",
                  "width": 34,
                },
                Object {
                  "align": "right",
                  "dataField": "subscriptionCount",
                  "headerClasses": "css-1lvmfbt",
                  "width": 33,
                },
                Object {
                  "dataField": "topicStatus",
                  "headerClasses": "css-1lvmfbt",
                  "text": "Status",
                  "width": 33,
                },
              ]
            }
            defaultHover={true}
            name="TopicList"
            onRowClick={[Function]}
            selectRow={
              Object {
                "bgColor": "#99ccff",
                "selected": undefined,
              }
            }
            uniqueKeyColumn="name"
          >
            <p
              className="emotion-3"
            >
              No data has been retrieved yet.
            </p>
          </DataTable>
        </TopicList>
      </div>
    </div>
  </div>
</TwoListDisplay>
`;

exports[`renders correctly before data is added 1`] = `
.emotion-3 {
  width: calc(50% - 30px);
  margin-left: 15px;
  margin-right: 15px;
  display: inline-block;
}

.emotion-7 {
  display: block;
}

.emotion-1 {
  display: block;
  height: 35px;
  margin: 2px;
}

.emotion-8 {
  min-width: 1000px;
}

.emotion-4 {
  border-left: 1px solid black;
  display: inline-block;
  height: calc(100% - 100px);
  position: fixed;
  z-index: -1;
}

.emotion-0.breadcrumb {
  float: left;
  height: 35px;
  margin-bottom: 1px;
}

.emotion-2 {
  text-align: center;
  float: left;
  width: 100%;
}

<div
  className="emotion-8"
>
  <div
    className="emotion-1"
  >
    <ol
      aria-label="breadcrumbs"
      className="emotion-0 breadcrumb"
      role="navigation"
    >
      <li
        className="active"
      >
        <span
          id="Home"
          onClick={[Function]}
        >
          Home
        </span>
      </li>
    </ol>
  </div>
  <div
    className="emotion-7"
  >
    <div
      className="emotion-3"
      id="left"
    >
      <div>
        <h2
          id="title"
        >
          Queues
        </h2>
      </div>
      <p
        className="emotion-2"
      >
        No data has been retrieved yet.
      </p>
    </div>
    <div
      className="emotion-4"
    />
    <div
      className="emotion-3"
      id="right"
    >
      <div>
        <h2
          id="title"
        >
          Topics
        </h2>
      </div>
      <p
        className="emotion-2"
      >
        No data has been retrieved yet.
      </p>
    </div>
  </div>
</div>
`;
